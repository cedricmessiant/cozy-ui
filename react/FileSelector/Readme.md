This component is used to display a customizable file selector.

### Default

```jsx
<FileSelector
  renderContent={() => <span>Click me to choose file</span>}
  processSelected={console.log}
/>
```

You render what you want:

```jsx
<FileSelector
  renderContent={() => <Icon icon="file" />}
  processSelected={console.log}
/>
```

### Multiple files

```jsx
<FileSelector
  renderContent={() => <span>Click me to choose files</span>}
  multiple
  processSelected={console.log}
/>
```

### Only accept images

```jsx
<FileSelector
  accept="image/*"
  renderContent={() => <span>Click me to choose an image</span>}
  multiple
  processSelected={console.log}
/>
```

### Simple input file

If you want a classic input file, just set `hidden` prop to `false`:

```jsx
<FileSelector hidden={false} processSelected={console.log} />
```
