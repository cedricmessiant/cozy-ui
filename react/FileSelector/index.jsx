import React from 'react'
import PropTypes from 'prop-types'

const FileSelector = ({
  className,
  disabled,
  hidden,
  multiple,
  processSelected,
  renderContent,
  ...inputProps
}) => (
  <label role="button" disabled={disabled} className={className}>
    {renderContent()}
    <input
      hidden={hidden}
      type="file"
      multiple={multiple}
      onChange={e => {
        if (multiple) {
          processSelected(Array.from(e.target.files))
        } else {
          processSelected(Array.from(e.target.files)[0])
        }
      }}
      {...inputProps}
    />
  </label>
)

FileSelector.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  hidden: PropTypes.bool,
  multiple: PropTypes.bool,
  processSelected: PropTypes.func.isRequired,
  renderContent: PropTypes.func
}

FileSelector.defaultProps = {
  disabled: false,
  hidden: true,
  multiple: false,
  renderContent: () => null
}

export default FileSelector
