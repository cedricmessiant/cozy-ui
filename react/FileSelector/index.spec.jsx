import React from 'react'

import FileSelector from './index'

describe('FileSelector component', () => {
  const processSelectedSpy = jest.fn()
  const pic1 = new File(['aaaa'], 'pic1.jpg')
  const pic2 = new File(['bbbb'], 'pic1.jpg')

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('should render a file selector', () => {
    const component = shallow(
      <FileSelector
        renderContent={() => <span>Click me</span>}
        processSelected={processSelectedSpy}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render a default file selector', () => {
    const component = shallow(
      <FileSelector hidden={false} processSelected={processSelectedSpy} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render a disabled file selector', () => {
    const component = shallow(
      <FileSelector
        disabled
        renderContent={() => <span>Click me</span>}
        processSelected={processSelectedSpy}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should pass props to the input', () => {
    const component = shallow(
      <FileSelector
        accept="image/*"
        renderContent={() => <span>Click me</span>}
        processSelected={processSelectedSpy}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should process selected file on change', () => {
    const filelist = [pic1]
    const component = shallow(
      <FileSelector
        accept="image/*"
        renderContent={() => <span>Click me</span>}
        processSelected={processSelectedSpy}
      />
    )
    component.find('input').simulate('change', {
      target: {
        files: filelist
      }
    })
    expect(processSelectedSpy).toHaveBeenCalledWith(pic1)
  })

  it('should process selected files on change if it is multiple', () => {
    const filelist = [pic1, pic2]
    const component = shallow(
      <FileSelector
        accept="image/*"
        multiple
        renderContent={() => <span>Click me</span>}
        processSelected={processSelectedSpy}
      />
    )
    component.find('input').simulate('change', {
      target: {
        files: filelist
      }
    })
    expect(processSelectedSpy).toHaveBeenCalledWith([pic1, pic2])
  })
})
