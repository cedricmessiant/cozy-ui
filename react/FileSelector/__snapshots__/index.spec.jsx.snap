// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSelector component should pass props to the input 1`] = `
<label
  disabled={false}
  role="button"
>
  <span>
    Click me
  </span>
  <input
    accept="image/*"
    hidden={true}
    multiple={false}
    onChange={[Function]}
    type="file"
  />
</label>
`;

exports[`FileSelector component should render a default file selector 1`] = `
<label
  disabled={false}
  role="button"
>
  <input
    hidden={false}
    multiple={false}
    onChange={[Function]}
    type="file"
  />
</label>
`;

exports[`FileSelector component should render a disabled file selector 1`] = `
<label
  disabled={true}
  role="button"
>
  <span>
    Click me
  </span>
  <input
    hidden={true}
    multiple={false}
    onChange={[Function]}
    type="file"
  />
</label>
`;

exports[`FileSelector component should render a file selector 1`] = `
<label
  disabled={false}
  role="button"
>
  <span>
    Click me
  </span>
  <input
    hidden={true}
    multiple={false}
    onChange={[Function]}
    type="file"
  />
</label>
`;
